[clang backend] In AArch64's DataLayout, specify a minimum function alignment of 4.

dougsonos opened the pull request:
This addresses an issue where the explicit alignment of 2 (for C++ ABI reasons) was being propagated to the back end and causing under-aligned functions (in special sections). (#90358)

This is an alternate approach suggested by @efriedma-quic in PR #90415.

~(There are a few failing tests. One I looked at was checking for the explicit alignment of 2, which no longer appears in AArch64 due to this change.)~

Commit:
In AArch64's DataLayout, specify a minimum function alignment of 4.
This addresses an issue where the explicit alignment of 2 (for C++
ABI reasons) was being propagated to the back end and causing
under-aligned functions (in special sections). (#90358)

efriedma-quic commented on llvm/lib/IR/AutoUpgrade.cpp:
```diff
@@ -5387,6 +5387,14 @@ std::string llvm::UpgradeDataLayoutString(StringRef DL, StringRef TT) {
+  if (T.isAArch64()) {
+    // Add "-Fn32"
+    if (!DL.contains("-Fn32"))
+      Res.append("-Fn32");
```
Ideally, the "fixed" datalayout string should be in the same order as the one a new compiler would generate (so if you're sticking it at the end here, stick it at the end on the actual strings).  I think we compare datalayouts using string equality in LTO.

dougsonos replied:
OK, thanks. I was afraid of that. I was copying two patterns - in ARM, `Fn8` seemed to be after any `p` and before `i`. In this file, the modifications are usually (always?) appending.

efriedma-quic replied:
Feel free to make the datalayout string whatever order is most convenient for the autoupgrade.

efriedma-quic reviewed (commented):
This looks correct.

I'd expect regression tests for the autoupgrade, and a test that clang isn't generating the alignment markings anymore.

Commit:
Move "-Fn32" to the end of the data layout strings, to make the
auto-upgrade simplest.

Commit:
Fix tests:
- CodeGen/target-data.c
- CodeGen/coff-aarch64-type-sizes.c
- CodeGen/aarch64-type-sizes.c
- CodeGen/CodeGenCXX/member-alignment.cpp
- OpenMP/distribute_parallel_for_num_threads_codegen.cpp

dougsonos commented:
> I'd expect regression tests for the autoupgrade, and a test that clang isn't generating the alignment markings anymore.

I just pushed:
- the `-Fn32` is at the end of the data layout strings
- the 6 tests that failed for me locally (testing only AArch64) are fixed

That includes a change to CodeGen/CodeGenCXX/member-alignment.cpp, which checks for alignment / no alignment on C++ methods, depending on target.

I'm not clear on when the autoupgrade runs and how to test it. Is there an example I can follow?

Thanks!

Commit:
AutoUpgrade: don't try to upgrade an empty string.
DataLayoutUpgradeTest: adapt test to this PR's change.

dougsonos commented:
Linux x86_64 has one test failure: RISCV/unnamed-sym-no-entry.c . AFAICT it looks unrelated.

dougsonos commented on llvm/unittests/Bitcode/DataLayoutUpgradeTest.cpp:
```diff
@@ -86,7 +86,7 @@ TEST(DataLayoutUpgradeTest, NoDataLayoutUpgrade) {
       "-f80:128:128-n8:16:32:64-S128");
   EXPECT_EQ(DL2, "e-m:e-i64:64-n32:64");
-  EXPECT_EQ(DL3, "e-m:o-i64:64-i128:128-n32:64-S128");
+  EXPECT_EQ(DL3, "e-m:o-i64:64-i128:128-n32:64-S128-Fn32");
```
Actually this is supposed to be a no-op test; this should be moved and replaced with one which tests a no-op situation.

Commit:
More correct DataLayoutUpgradeTests

fhahn commented on clang/lib/Basic/Targets/AArch64.cpp:
```diff
@@ -1480,11 +1480,11 @@ AArch64leTargetInfo::AArch64leTargetInfo(const llvm::Triple &Triple,
   if (getTriple().isOSBinFormatMachO()) {
     if(getTriple().isArch32Bit())
-      resetDataLayout("e-m:o-p:32:32-i64:64-i128:128-n32:64-S128", "_");
+      resetDataLayout("e-m:o-p:32:32-i64:64-i128:128-n32:64-S128-Fn32", "_");
```
Just to double check @TNorthover, is an alignment of 32 OK for arm64_32?

TNorthover replied:
Yep. CPU would hard-fault if that was violated. All instructions are 4 bytes and must be aligned.

AtariDreams replied:
Is this true even if the cpu was executing thumb2 instructions in AArch32 mode?

efriedma-quic replied:
Thumb2 function pointers aren't aligned... but from LLVM's perspective it's a completely different target, so it's not relevant.

fhahn reviewed (approved):
LGTM, thanks!

fhahn committed:
Merge branch 'main' into min_sect_align_2

efriedma-quic reviewed (approved):
LGTM

efriedma-quic commented:
Which email address do you want the commit associated with?  The GitHub "squash and merge" button wants to attribute it to your "sonosphere.com" address, but it looks like the commits themselves are using an "apple.com" address.

AtariDreams commented:
/cherry-pick ddecadabebdd4b301bd65534b58009e57ac1bbe5

AtariDreams commented:
/cherry-pick ddecadabebdd4b301bd65534b58009e57ac1bbe5
